pkg/controller/machinepool/machinepool_controller.go

```diff
@@ -617,25 +640,35 @@ func (r *ReconcileMachinePool) syncMachineSets(
 				if rMS.Spec.Template.Spec.Taints == nil {
 					rMS.Spec.Template.Spec.Taints = []corev1.Taint{}
 				}
-				// Make a temporary map of generated MachineSet's taints for easy lookup
-				gTaintsByKey := make(map[string]*corev1.Taint)
-				for gIndex, gTaint := range ms.Spec.Template.Spec.Taints {
-					gTaintsByKey[gTaint.Key] = &ms.Spec.Template.Spec.Taints[gIndex]
+				// First, build a list of remote MachineSet taints that do not have taints needing to be deleted
+				taints := make([]corev1.Taint, 0)
+				for _, rTaint := range rMS.Spec.Template.Spec.Taints {
+					if taintsToDelete.Has(identifierForTaint(&rTaint)) {
+						// This entry would be deleted from the final list
+						objectModified = true
+						continue
+					}
+					taints = append(taints, rTaint)
 				}
-				// Go through the remote MachineSet's taints, replacing overlaps with the generated MachineSet
-				for rIndex, rTaint := range rMS.Spec.Template.Spec.Taints {
-					if gTaint, foundTaint := gTaintsByKey[rTaint.Key]; foundTaint {
-						if gTaint.Value != rTaint.Value || gTaint.Effect != rTaint.Effect {
-							rMS.Spec.Template.Spec.Taints[rIndex] = *gTaint
-							objectModified = true
+				rMS.Spec.Template.Spec.Taints = taints
+
+				for _, gTaint := range ms.Spec.Template.Spec.Taints {
+					// Taints are provided as a list, so duplicate entries are possible.
+					// Use a flag to indicate if a taint was found, append if not.
+					foundTaint := false
+					for rIndex, rTaint := range rMS.Spec.Template.Spec.Taints {
+						if reflect.DeepEqual(identifierForTaint(&gTaint), identifierForTaint(&rTaint)) {
+							foundTaint = true
+							if gTaint.Value != rTaint.Value {
+								rMS.Spec.Template.Spec.Taints[rIndex].Value = gTaint.Value
+								objectModified = true
+							}
 						}
-						delete(gTaintsByKey, rTaint.Key)
```

Comment:
@2uasimojo Your code for taints was sophisticated but it still had the fatal flaw of not being able to handle duplicate entries for taints. I tried fixing it but it kept on getting too complicated, so I ditched that and opted for keeping track with a flag approach. It does require looping twice, but it is similar to the handling of labels, hence easier to read